syft/frameworks/torch/mpc/fss.py

```diff
@@ -195,210 +226,365 @@ def le(x1, x2):
 class DPF:
     """Distributed Point Function - used for equality"""
 
-    def __init__(self):
-        pass
-
     @staticmethod
     def keygen(n_values=1):
-        beta = th.tensor([1], dtype=dtype)
-        alpha = th.randint(0, 2 ** n, (n_values,))
-
+        alpha = np.random.randint(0, 2 ** n, size=(n_values,), dtype=np.uint64)
+        beta = np.array([1])
         α = bit_decomposition(alpha)
         s, t, CW = (
-            Array(n + 1, 2, λ, n_values),
+            Array(n + 1, 2, λs, n_values),
             Array(n + 1, 2, n_values),
-            Array(n, 2 * (λ + 1), n_values),
+            Array(n, 2, (λs + 1), n_values),
         )
-        s[0] = randbit(size=(2, λ, n_values))
-        t[0] = th.tensor([[0, 1]] * n_values, dtype=th.uint8).t()
+        _CW = []
+        s[0] = randbit(shape=(2, λ, n_values))
+        t[0] = np.array([[0, 1]] * n_values).T
         for i in range(0, n):
             g0 = G(s[i, 0])
             g1 = G(s[i, 1])
             # Re-use useless randomness
-            sL_0, _, sR_0, _ = split(g0, [λ, 1, λ, 1])
-            sL_1, _, sR_1, _ = split(g1, [λ, 1, λ, 1])
+            sL_0, _, sR_0, _ = split(g0, (EQ, λs, 1, λs, 1))
+            sL_1, _, sR_1, _ = split(g1, (EQ, λs, 1, λs, 1))
             s_rand = (sL_0 ^ sL_1) * α[i] + (sR_0 ^ sR_1) * (1 - α[i])
 
-            cw_i = TruthTableDPF(s_rand, α[i])
+            cw_i = SwitchTableDPF(s_rand, α[i])
             CW[i] = cw_i ^ g0 ^ g1
+            _CW.append(compress(CW[i], α[i], op=EQ))
+            CWi = uncompress(_CW[i])
 
             for b in (0, 1):
-                τ = [g0, g1][b] ^ (t[i, b] * CW[i])
-                τ = τ.reshape(2, λ + 1, n_values)
-                # filtered_τ = τ[𝛼[i]] OLD
-                α_i = α[i].unsqueeze(0).expand(λ + 1, n_values).unsqueeze(0).long()
-                filtered_τ = th.gather(τ, 0, α_i).squeeze(0)
-                s[i + 1, b], t[i + 1, b] = split(filtered_τ, [λ, 1])
+                τ = [g0, g1][b] ^ (t[i, b] * CWi)
```

Comment:
Is there a difference between those two t's, because it's kind of hard to spot the difference at first.

Reply:
Between t and τ (tau)?

Reply:
Or the two lines of code? (in which case `CW[i]` -> `CWi`

Reply:
Between t and τ (tau), but I guess it make more sense when having the paper in hand and reading the code

Reply:
tau here is an intermediate variable (that I call `state` in the DIF btw, I'll rename it as such) which is used to select the first part or second part of the 2(λ + 1) bit-long `[g0, g1][b] ^ (t[i, b] * CWi)` depending of the bit `α[i]` .
Related to the paper: it consists of choosing the left branch or the right branch when descending the tree